### DISCUSSION TITLE
Print on a BLE printer

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#734] [Q&A]
Good evening,
I'm trying to print on a BLE printer. 
I'm using the example "https://github.com/Moddable-OpenSource/moddable/blob/public/examples/network/ble/uart-client/main.js"
The connection  and "onServices" event are done, but no "onCharacteristics(" event is received.
I tried to change "service" and "characteristics" UUID, with 
"
const SERVICE_UUID = uuid`38eb4a80-c570-11e3-9507-0002a5d5c51b`;
const CHARACTERISTIC_RX_UUID = uuid`38eb4a81-c570-11e3-9507-0002a5d5c51b`;
const CHARACTERISTIC_TX_UUID = uuid`38eb4a82-c570-11e3-9507-0002a5d5c51b`;
"
but it doesn't work.
What can I do?
regards

**Comment by mauroForlimpopoli** [comment]
...the printer has a POS/ESC command

**Comment by mauroForlimpopoli** [comment]
..partial correction...
I receive the onServices" event but the array is empty.
I found some example on the web and it seems that it simplest than I expect. It could be enough to send a text string with ESC/POS command after the connection to set and print data. Is it possible, how can I do that ?
Regards

**Comment by mauroForlimpopoli** [comment]
when I try to write to a server I always get this error.

![Immagine3](https://user-images.githubusercontent.com/78380771/143973357-e82e8974-fb8e-419a-99ba-3ea6996ff4f7.png)
In the array which Arg is wrong?

**Comment by phoddie** [comment]
The second argument must be an `ArrayBuffer`. You are passing a string. Use `ArrayBuffer.fromString()` to convert the string to an `ArrayBuffer` before making the call.

**Reply by mauroForlimpopoli** [reply]
I done the correction and now it works !!!
Thank you very much !

**Reply by phoddie** [reply]
@mauroForlimpopoli - I'm curious. Were you able to get your BLE printer working?

**Reply by mauroForlimpopoli** [reply]
Yes !!! 
for first I found the "name" of the printer (it was "Printer Bluetooth")
then this is shorty my code:
```
	const SERVICE_UUID = uuid`00001800-0000-1000-8000-00805f9b34fb`;
	const CHARACTERISTIC_RX_UUID = uuid`49535343-1E4D-4BD9-BA61-23C647249616`;
	const CHARACTERISTIC_TX_UUID = uuid`49535343-8841-43F4-A8D4-ECBE34729BB3`;

```
This is the main step for the connection 
```
		onServices(services) {	
			services.forEach(service => {
				let uuids = service.uuid;
			});		
			if (services.length)
				services[3].discoverAllCharacteristics();
		}

```
Of course I use the printer only using the POS standard protocol
